include/pybind11/stl_bind.h

```diff
@@ -352,10 +352,10 @@ void vector_accessor(enable_if_t<vector_needs_copy<Vector>::value, Class_> &cl)
     using DiffType = typename Vector::difference_type;
     using ItType = typename Vector::iterator;
     cl.def("__getitem__", [](const Vector &v, DiffType i) -> T {
-        if (i < 0 && (i += v.size()) < 0) {
-            throw index_error();
+        if (i < 0) {
+            i += v.size();
         }
-        if ((SizeType) i >= v.size()) {
+        if (i < 0 || (SizeType) i >= v.size()) {
```

Comment:
This is microscopically slower than it was before since now it does the i<0 check twice even if ti doesn't need to, but it seems a bit more readable now. Also, if we do care about performance in this case, we should fix the wrap_i function as well.

Reply:
This is mostly to make it "nice" (subjectively), I'm really not worried about performance one way or another (untested):
```
if (i < 0) {
    i += v.size();
    if (i < 0) {
        throw index_errror();
    }
}
auto i_st = static_cast<SizeType>(i);
if (i_st >= v.size()) {
    throw index_error();
}
return v[i_st];
```

What's missing here IMO is nice error reporting (e.g. something like "index -8 is out of range for array size 5"), but that's beyond the scope of this PR.